### PR TITLE
Sockets - Jansen

### PR DESCRIPTION
**Pull request by JansenMartin**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | I had a tendency to make a single class do too much.  For example, I originally had my 'is_available?' method inside of Manager.  After reading POODR chapter 3, though, I realized that functionality would be better suited for Rooms.
What was a design decision you made that changed over time over the project? | I originally had Rooms and Reservation mainly exist to make instances of themselves.  Over time, though, I gradually added more functionality that was originally supposed to go under Manager (such as 'is_available?' for Rooms and 'total_cost' for Reservation).
What was a concept you gained clarity on, or a learning that you'd like to share? | While reading about class dependencies in POODR, I saw the benefit of having clear divides between what classes do and don't know.  Too much overlap makes it more difficult to debug when you inevitably refactor your code.
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | My tests for the 'is_available?' method check a variety of check-in and check-out date combinations.  Making sure a booking date that STARTS BEFORE and ENDS AFTER an existing reservation returns FALSE would be a **nominal case**, because there is a clear, obvious overlap between the two.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? |  The 'validate_date' method in Reservation ensures that the user gives a viable date range.  My tests cover what would happen if someone tries to enter the date ranges in MM/DD/YYYY or YYYY/DD/MM formats.  It's an **edge case** because these formats could (and should) trigger errors since the code is written under the assumption that we're using a YYYY/MM/DD format.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | When starting from scratch, this was extremely helpful in pushing past the initial block that comes from looking at a blank canvas.  As I got deeper into the project, however, it was easier for me to forget to write my tests first.  The pseudocode/tests/code pattern definitely made the beginning stages smoother than other parts of the project -- That's a lesson I'll carry with me in the future.

### CODE REVIEW

File: lib/manager.rb [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+require_relative "reservation.rb"
+
+module Hotel
+  class Manager
+    attr_reader :rooms, :reservations
+
+    def initialize
+      @rooms = (1..20).map do |room_number|
+        Hotel::Room.new(room_number: room_number)
+      end
```
Comment (droberts-sea) [inline lib/manager.rb]: Since we're inside the `Hotel` module already, you don't need the `Hotel::` here.

File: lib/manager.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+      @reservations = []
+    end
+
+    def reserve_room(check_in, check_out, room_selection: nil)
+      available_room_list = list_available_rooms(check_in, check_out)
+
+      if available_room_list == []
+        return available_room_list
```
Comment (droberts-sea) [inline lib/manager.rb:21]: I'm not sure that `return`ing an empty array is the right approach here. You might consider either raising an exception or returning `nil`.

File: lib/manager.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+      elsif room_selection != nil
+        available_room = ""
+        available_room_list.each do |room|
+          available_room = room if room.room_number == room_selection
+        end
+        if available_room == ""
```
Comment (droberts-sea) [inline lib/manager.rb:27]: I think it might be possible to tighten up this code a little.

File: lib/manager.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    end
+
+    def list_available_rooms(check_in, check_out)
+      list = []
+      booking_range = (Date.parse(check_in)..Date.parse(check_out))
+
+      @rooms.each do |room|
+        if room.is_available?(booking_range)
+          list << room
+        end
```
Comment (droberts-sea) [inline lib/manager.rb:50]: This code shows me that you've done a good job breaking up functionality between classes. You could imagine having `Manager` do a bunch of date math here, but instead it asks the room whether it's available. This is what Metz is talking about when she says you should "ask for _what_ instead of telling _how_".

File: lib/reservation.rb [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+      @total_cost = total_cost(check_in, check_out, room.cost_per_night)
+      @id = id
+
+      validate_date
+    end
+
+    def total_cost(check_in, check_out, cost_per_night)
+      number_of_nights = number_of_nights(check_in, check_out)
+      total_cost = cost_per_night * number_of_nights
+
```
Comment (droberts-sea) [inline lib/reservation.rb:24]: I like the decision to make this a method rather than storing the cost as an instance variable. That way if the number of nights or the cost per night were to change (say in a future version of the program), the total cost will automatically be correct.

File: lib/manager.rb [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+    end
+
+    def list_reservations_on(date)
+      list = []
+      date = Date.parse(date)
+
+      @reservations.each do |reservation|
+        check_in = reservation.check_in
+        check_out = reservation.check_out
+
```
Comment (droberts-sea) [inline lib/manager.rb:63]: The `.select` enumerable might be helpful to clean up this code.

File: lib/manager.rb [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+        if (check_in..check_out).include?(date)
+          list << reservation
+        end
+      end
+    end
```
Comment (droberts-sea) [inline lib/manager.rb:70]: This method has a bug! You do all this work in the each loop, but you don't `return list` at the end. So, you end up implicitly returning the result of `.each`, which is the original array.

```ruby
$ pry -r ./lib/manager.rb
pry> m = Hotel::Manager.new
pry> m.reserve_room('2019-04-07', '2019-04-11')
pry> m.list_reservations_on('2020-01-01')
=> [#<Hotel::Reservation:0x00007feaa1922280
  @check_in=#<Date: 2019-04-07 ((2458581j,0s,0n),+0s,2299161j)>,
  @check_out=#<Date: 2019-04-11 ((2458585j,0s,0n),+0s,2299161j)>,
  @id=1,
  @room=#<Hotel::Room:0x00007feaa10dcd28>]
```

Comment (droberts-sea) [reply lib/manager.rb:70]: This is something that your tests should have caught.

File: spec/manager_spec.rb [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    end
+
+    it "allows someone to reserve a specific room" do
+      reservation = @manager.reserve_room("2019-3-20", "2019-3-21", room_selection: 3)
+      expect(reservation.room.room_number).must_equal 3
+    end
+
+    it "raises an ArgumentError if someone tries to reserve a specific room that's unavailable" do
+      expect { @manager.reserve_room("2019-3-20", "2019-3-21", room_selection: 1) }.must_raise ArgumentError
+    end
```
Comment (droberts-sea) [inline spec/manager_spec.rb:70]: What if you try to reserve without a specific room, and the whole hotel is booked?

File: spec/manager_spec.rb [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+  end
+
+  describe "list_reservations_on method" do
+    it "can list reservations associated with a specific date" do
+      5.times do
+        reservation = @manager.reserve_room("2019-3-20", "2019-3-25")
```
Comment (droberts-sea) [inline spec/manager_spec.rb]: What about reservations that _aren't_ associated with that date?

File: spec/room_spec.rb [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+      date_range = (check_in..check_out)
+      availability = @reserved_room.is_available?(date_range)
+
+      expect(availability).must_equal false
+    end
+
+    it "returns false if the given date range STARTS on reservaton's check_in day and ENDS DURING" do
+      check_in = Date.parse(@check_in)
+      check_out = Date.parse("2020-03-22")
+      date_range = (check_in..check_out)
```
Comment (droberts-sea) [inline spec/room_spec.rb:60]: Good work laying out all of these different test cases.

Since you recreate them so frequently, it might make sense to store the instances of `Date` as instance variables in the `before` block as well. Then in the tests, you could do something like:

```ruby
date_range = @check_in_date..(@check_out_date - 2)
```

**Comment by droberts-sea** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | yes
Answer comprehension questions | yes
**Design** |
Each class is responsible for a single piece of the program | yes
Classes are loosely coupled | yes
**Wave 1** |
List rooms | yes
Reserve a room for a given date range | yes
List reservations for a given date | yes
Calculate reservation price | yes
Invalid date range produces an error | yes
Test coverage | mostly - see inline
**Wave 2** |
View available rooms for a given date range | yes
Reserving a room that is not available produces an error | yes
Test coverage | yes
**Wave 3** |
Create a block of rooms | N/A
Check if a block has rooms | N/A
Reserve a room from a block | N/A
Test coverage | N/A
**Fundamentals** |
Names variables, classes and modules appropriately | yes
Understanding of variable scope - local vs instance | yes
Can create complex logical structures utilizing variables | yes
Appropriately uses methods to break down tasks into smaller simpler tasks | yes
Understands the differences between class and instance methods | yes
Appropriately uses iterators and `Enumerable` methods | yes
Appropriately writes and utilizes classes | yes
Appropriately utilizes modules as a namespace | yes
**Wrap Up** |
There is a refactors.txt file | yes
The file provides a roadmap to future changes | yes
**Additional Feedback** | Great job overall! This code is well-organized, well-written and mostly well-tested. The design you ended up with does a good job of splitting responsibilities up into classes, and is easy to understand. I've left a number of inline comments below for you to review, but in general I like what I see. Keep up the hard work!
